### PR TITLE
add tinycthreadpool

### PR DESCRIPTION
**Pull request by bennyhuo**
Specify library name and version:  **tinycthreadpool/1.0**

This is a portable C threadpool. It is based on [tinycthread](https://github.com/tinycthread/tinycthread) and supports both Windows and POSIX systems.
---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/tinycthreadpool/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import os
+
+from conan import ConanFile
+from conans import CMake, tools
+from conan.tools.files import get
+
+required_conan_version = ">=1.53.0"
+
+#
+# INFO: Please, remove all comments before pushing your PR!
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:10]: @bennyhuo Please, follow the comment indication.

File: recipes/tinycthreadpool/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#
+
+
+class PackageConan(ConanFile):
+    name = "tinycthreadpool"
+    version = "1.0"
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:16]:
```suggestion
```
No version attribute, so we can re-use the same recipe.

File: recipes/tinycthreadpool/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    description = "Portable C thread pool"
+    # Use short name only, conform to SPDX License List: https://spdx.org/licenses/
+    # In case not listed there, use "LicenseRef-<license-file-name>"
+    license = "https://github.com/bennyhuo/tinycthreadpool/blob/master/LICENSE"
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:20]: Use the SPDX short-name only, as described in the comment.

File: recipes/tinycthreadpool/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/bennyhuo/tinycthreadpool"
+    # no "conan" and project name in topics. Use topics from the upstream listed on GH
+    topics = ("C", "Thread Pool", "tinycthread")
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:24]:
```suggestion
    topics = ("thread-pool", "threading", "pure-c")
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
+    default_options = {"shared": False}
+    generators = "cmake"
+    
+    # no exports_sources attribute, but export_sources(self) method instead
+    # this allows finer grain exportation of patches per version
+    def export_sources(self):
+        self.copy(os.path.join("src/*"))
+
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:34]:
```suggestion
```
There is nothing to be exported here in this PR.

File: recipes/tinycthreadpool/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    # In case not listed there, use "LicenseRef-<license-file-name>"
+    license = "https://github.com/bennyhuo/tinycthreadpool/blob/master/LICENSE"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/bennyhuo/tinycthreadpool"
+    # no "conan" and project name in topics. Use topics from the upstream listed on GH
+    topics = ("C", "Thread Pool", "tinycthread")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
+    default_options = {"shared": False}
+    generators = "cmake"
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:28]:
```suggestion
    generators = "cmake"
```
Consider migrating it to Conan 2.0. The `cmake` generator is outdated and will become deprecated very soon. https://docs.conan.io/en/latest/conan_v2.html

File: recipes/tinycthreadpool/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        self.copy("*.lib", dst="lib", keep_path=False)
+        self.copy("*.dll", dst="bin", keep_path=False)
+        self.copy("*.dylib*", dst="lib", keep_path=False)
+        self.copy("*.so", dst="lib", keep_path=False)
+        self.copy("*.a", dst="lib", keep_path=False)
+        self.copy("LICENSE", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
+
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:61]:
```suggestion
        self.cpp_info.libs = ["tinycthreadpool"]
```
Avoid collect_libs, it's fragile and does not respect link order. Also, keeping it explicit is much easier than deduce what library name will be available.

File: recipes/tinycthreadpool/all/test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 2.8.12)
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/test_package/CMakeLists.txt:1]: Please, use the template from https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package/all/test_package

CONAN_LIBS is no longer used and it's almost deprecated. Also, consider only Cmake targets for usage.

**Review by uilianries — COMMENTED** [review]
Thank you for your contribution. I see you are using Conan in your project already 😸

File: recipes/tinycthreadpool/all/conanfile.py (hunk 6 above)
Comment (bennyhuo) [reply recipes/tinycthreadpool/all/conanfile.py:28]: It seems that the 'generators' is needed for conan 1.x, right?

Comment (bennyhuo) [reply recipes/tinycthreadpool/all/conanfile.py:28]: If I simply remove this line, it won't work. And I don't know how to make it compile. Maybe we should move on to conan 2.0 later on.

Comment (bennyhuo) [reply recipes/tinycthreadpool/all/conanfile.py]: Ah, I see. I've just changed to 'CMakeDeps'.

**Comment by bennyhuo** [comment]
I have updated the recipe to make all the checks passed. Please let me know if there is something else to do to meet the requirements. Thanks.

**Comment by bennyhuo** [comment]
@uilianries Hi there, I just wonder when can we have this pull request merged? Are there any problems I need to fix? : )

File: recipes/tinycthreadpool/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import os
+
+from conan import ConanFile
+from conans import CMake, tools
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:4]:
```suggestion
from conan.tools.cmake import CMake
```
Please, avoid mixing `conans` and `conan`. With Conan 2.x, `conans` will be deprecated, so we are migrating everything.

Also, consider CMakeToolchain generator, much better that only workaround with `conan_basic_setup()` injected in your cmake file.


Consider https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package  as example

File: recipes/tinycthreadpool/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+class TinyCThreadPoolConan(ConanFile):
+    name = "tinycthreadpool"
+    description = "Portable C thread pool"
+    license = "BSD-2-Clause"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/bennyhuo/tinycthreadpool"
+    topics = ("thread-pool", "threading", "pure-c")
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "CMakeDeps"
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:20]:
```suggestion
    generators = "CMakeDeps", "CMakeToolchain"
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure(source_folder="src")
+        cmake.build()
+
+    def package(self):
+        self.copy("*.h", dst="include", src="src")
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:35]: As you are the author of that project, you could consider add `install()` support to your cmake file, so you don't need to copy pattern by pattern 😉

File: recipes/tinycthreadpool/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        self.cpp_info.libs = ["tinycthreadpool"]
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs = ["pthread"]
+
+    def validate(self):
+        if self.settings.compiler == "Visual Studio":
+            if Version(self.settings.compiler.version) < "16":
+                raise ConanInvalidConfiguration(f"TinyCThreadPool doesn't support MSVC version: {self.settings.compiler.version}")
+            if self.settings.compiler.runtime in ['MTd', 'MT']:
+                raise ConanInvalidConfiguration(f"TinyCThreadPool doesn't support this runtime {self.settings.compiler.runtime}")
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:53]:
```suggestion
```
Let's move it to the correct order, https://docs.conan.io/en/latest/reference/commands/creator/create.html#methods-execution-order

File: recipes/tinycthreadpool/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/bennyhuo/tinycthreadpool"
+    topics = ("thread-pool", "threading", "pure-c")
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "CMakeDeps"
+    
+    def configure(self):
+        self.settings.rm_safe("compiler.libcxx")
+        self.settings.rm_safe("compiler.cppstd")
+
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/conanfile.py:25]:
```suggestion

    def validate(self):
        check_min_vs(self, "1921")
        if is_msvc_static_runtime(self):
            raise ConanInvalidConfiguration("{self.ref} does not support static runtime.")

```
We have new and powerful tool for Microsoft https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html

File: recipes/tinycthreadpool/all/test_package/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.15)
```
Comment (uilianries) [inline recipes/tinycthreadpool/all/test_package/CMakeLists.txt:1]: Please, follow https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package/all/test_package template, and please, add test_v1_package too. You can copy directly from the template and update.

CONAN_LIBS is almost deprecated, we no longer recommend as it's fragile. Instead, prefer CMake targets, which is safer for link and finding folders.

**Review by uilianries — COMMENTED** [review]
@bennyhuo  Thank you for your contribution! Please, take a look on my review

**Comment by uilianries** [comment]
@bennyhuo no fPIC neither shared option support? Does tinycthreadpool only supports static library?

**Comment by bennyhuo** [comment]
@uilianries Hi there! I've managed to rewrite the recipes but still get two errors.

The first one is on Windows with msvc. When building a shared library, the *.dll file is installed into the 'bin' folder, and the error is:

```
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component tinycthreadpool::tinycthreadpool library 'tinycthreadpool' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054-LIBRARY-DOES-NOT-EXIST) 
```

I've no idea what the problem is. Maybe I should configure to generate a import library for this?

And the second one is on macOS. It seems the error is related to the @rpath. 

```
dyld: Library not loaded: libtinycthreadpool.dylib
  Referenced from: /Users/jenkins/w/prod/BuildSingleReference@8/cci-6dc17ea4/recipes/tinycthreadpool/all/test_package/build/Debug/./example
  Reason: image not found
tinycthreadpool/1.0 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
tinycthreadpool/1.0 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
ERROR: tinycthreadpool/1.0 (test package): Error in test() method, line 26
	self.run(cmd, env="conanrun")
	ConanException: Error -6 while executing ./example
```

Do you have some advice for this two problems? Thank you in advance.

**Comment by bennyhuo** [comment]
@uilianries  The error reported from conan v1 pipeline indicates that the conan v1 pipeline run the test_package instead of test_v1_package, is that designed to be?

When building the library as shared, `run_environment=True` is needed to call the `self.run(cmd, env="conanrun")` in Conan V1 Test to find the `*.dll` or `*.dylib` files but there is no such argument (also not necessary) in Conan V2. 

Maybe I am still missing something?

```
/Users/jenkins/w/prod/BuildSingleReference@7/cci-00355e70/recipes/tinycthreadpool/all/test_package/example.c:8:40: warning: format specifies type 'unsigned int' but the argument has type 'threadpool_t *' (aka 'struct threadpool_t *') [-Wformat]
  printf("Create thread pool: %#x.\n", threadpool);
                              ~~~      ^~~~~~~~~~
1 warning generated.
dyld: Library not loaded: libtinycthreadpool.dylib
  Referenced from: /Users/jenkins/w/prod/BuildSingleReference@7/cci-00355e70/recipes/tinycthreadpool/all/test_package/build/Debug/./example
  Reason: image not found
tinycthreadpool/1.0 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
tinycthreadpool/1.0 (test package): WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
ERROR: tinycthreadpool/1.0 (test package): Error in test() method, line 26
	self.run(cmd, env="conanrun")
	ConanException: Error -6 while executing ./example
```

**Comment by bennyhuo** [comment]
@uilianries I have managed to pass all the checks.

**Comment by bennyhuo** [comment]
@uilianries Hello, is there any news for this pr?

**Comment by prince-chrismc** [comment]
👋  Uillian is on vacation hence why we has not followed up. But I should have follow up... my bot track all the PRs but this was not on the list https://github.com/prince-chrismc/conan-center-index-pending-review/issues/1 

I found a bug in my code :)

****


image

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

File: recipes/tinycthreadpool/all/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
+import os
+
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:4]: Linter flagged as unused

```suggestion
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+from conan.tools.microsoft import is_msvc_static_runtime, is_msvc
+from conan.tools.scm import Version
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:6]: Linter

```suggestion
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+            self.options.rm_safe("fPIC")
+        self.settings.rm_safe("compiler.libcxx")
+        self.settings.rm_safe("compiler.cppstd")
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def source(self):
+        get(self, **self.conan_data["sources"][self.version], destination=self.source_folder, strip_root=True)
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:41]: This is the default so we can skip it... less is more

```suggestion
        get(self, **self.conan_data["sources"][self.version], strip_root=True)
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        if is_msvc(self):
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
+
+        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:51]: This was fixed in 1.54

```suggestion
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
+import os
+
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.microsoft import is_msvc_static_runtime, is_msvc
+from conan.tools.scm import Version
+from conan.tools.files import get, copy
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+
+required_conan_version = ">=1.53.0"
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:10]: improved toolchain to skip a policy

```suggestion
required_conan_version = ">=1.54.0"
```

File: recipes/tinycthreadpool/all/conanfile.py [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        self.cpp_info.set_property("pkg_config_name", "tinycthreadpool")
+
+        self.cpp_info.libs = ["tinycthreadpool"]
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("pthread")
+
+        # TODO: to remove in conan v2
+        self.cpp_info.names["cmake_find_package"] = "tinycthreadpool"
+        self.cpp_info.names["cmake_find_package_multi"] = "tinycthreadpool"
+
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:79]: Looking at https://github.com/bennyhuo/tinycthreadpool/blob/master/CMakeLists.txt there are no config files being installed so we do not need to list them here

The values being set are the same as the defaults so nothing else should change :)

```suggestion
    def package_info(self):
        self.cpp_info.libs = ["tinycthreadpool"]
        if self.settings.os in ["Linux", "FreeBSD"]:
            self.cpp_info.system_libs.append("pthread")
```

**Review by prince-chrismc — COMMENTED** [review]
Overall looks good, since this has been open for a while (sorry about that) there's more updates I'd like to sneak in

**Review by prince-chrismc — COMMENTED** [review]
I have some concerns about this project, since it's a fork we usually do not allow those

and when they add enough value we ask for teh author prefix... but when I was looking into I noticed this has a copy of 

https://github.com/tinycthread/tinycthread

embeeded inside ... which is fine but for conan that  would ideal be a dependency since people could want to use two lirbaries that depend on it this and something else that would hit linkage issues.

Have you considered linking against  this library instead of copying it?

https://github.com/mbrossard/threadpool/compare/master...bennyhuo:tinycthreadpool:master#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Comment by SSE4** [comment]
> I found a bug in my code :)

so what's the bug here?

**Comment by bennyhuo** [comment]
> I have some concerns about this project, since it's a fork we usually do not allow those
> 
> and when they add enough value we ask for teh author prefix... but when I was looking into I noticed this has a copy of
> 
> https://github.com/tinycthread/tinycthread
> 
> embeeded inside ... which is fine but for conan that would ideal be a dependency since people could want to use two lirbaries that depend on it this and something else that would hit linkage issues.
> 
> Have you considered linking against this library instead of copying it?
> 
> [mbrossard/threadpool@master...bennyhuo:tinycthreadpool:master#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5](https://github.com/mbrossard/threadpool/compare/master...bennyhuo:tinycthreadpool:master#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)

I have tried, but the recipe does not work well with the new conan version.

```
tinycthread/cci.20161001: Downloaded recipe revision bbbf04729d59be014078ab872ab2482f
Graph root
    conanfile.txt: /Users/benny/WorkSpace/external/tinycthreadpool/conanfile.txt
*********************************************************
Recipe 'tinycthread/cci.20161001' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'tinycthread/cci.20161001' not resolved: tinycthread/cci.20161001: Cannot load recipe.
Error loading conanfile at '/Users/benny/.conan2/p/tinyc23a3a45c6c35e/e/conanfile.py': Unable to load conanfile in /Users/benny/.conan2/p/tinyc23a3a45c6c35e/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Users/benny/.conan2/p/tinyc23a3a45c6c35e/e/conanfile.py", line 1, in <module>
    from conans import CMake, ConanFile, tools
ImportError: cannot import name 'CMake' from 'conans' (/Users/benny/Library/Python/3.9/lib/python/site-packages/conans/__init__.py)
```

**Comment by bennyhuo** [comment]
Hello, any news about this? @prince-chrismc

**Comment by prince-chrismc** [comment]
I'd suggest updateting you build script to link against the original library instead of trying to duplicate it, I made #17001 which is pretty small and will support v2

but we can make sure everything works with v1 in the mean time

**Comment by bennyhuo** [comment]
> I'd suggest updateting you build script to link against the original library instead of trying to duplicate it, I made #17001 which is pretty small and will support v2
> 
> but we can make sure everything works with v1 in the mean time

Great. I shall update my project after  #17001 merged. Thanks.

**Comment by prince-chrismc** [comment]
It was merged, drop a comment here if you have any questions. you can also find me in the cppalaince slack's Conan channel :)

**Comment by bennyhuo** [comment]
> It was merged, drop a comment here if you have any questions. you can also find me in the cppalaince slack's Conan channel :)

Great news, I will follow this up asap.

**Comment by bennyhuo** [comment]
> ce

Done.

**Review by prince-chrismc — APPROVED** [review]

File: recipes/tinycthreadpool/all/conanfile.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+    def layout(self):
+        cmake_layout(self, src_folder="src")
+
+    def requirements(self):
+        self.requires("tinycthread/cci.20161001", transitive_headers=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        if is_msvc(self):
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
```
Comment (prince-chrismc) [inline recipes/tinycthreadpool/all/conanfile.py:50]: https://github.com/search?q=repo%3Abennyhuo%2Ftinycthreadpool%20USE_MSVC_RUNTIME_LIBRARY_DLL&type=code

I do not see this in your project

```suggestion
```

**Comment by prince-chrismc** [comment]
:) 🤞

**Review by danimtb — APPROVED** [review]

**Pull request merged**